modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(DATA_LIFECYCLE_JOB_NAME)) {
+            logger.trace("DLM job triggered: " + event.getJobName() + ", " + event.getScheduledTime() + ", " + event.getTriggeredTime());
+            run(clusterService.state());
+        }
+    }
+
+    /**
+     * Iterates over the DLM managed data streams and executes the needed operations
+     * to satisfy the configured {@link org.elasticsearch.cluster.metadata.DataLifecycle}.
+     */
+    // default visibility for testing purposes
+    void run(ClusterState state) {
+        for (DataStream dataStream : state.metadata().dataStreams().values()) {
+            if (dataStream.getLifecycle() == null) {
+                continue;
+            }
+
+            IndexMetadata writeIndex = state.metadata().index(dataStream.getWriteIndex());
+            if (writeIndex != null && isManagedByDLM(dataStream, writeIndex)) {
+                RolloverRequest rolloverRequest = defaultRolloverRequestSupplier.apply(dataStream.getName());
+                transportActionsDeduplicator.executeOnce(
+                    rolloverRequest,
+                    ActionListener.noop(),
+                    (req, reqListener) -> rolloverDataStream(rolloverRequest, reqListener)
+                );
+            }
+
+            TimeValue retention = getRetentionConfiguration(dataStream);
+            if (retention != null) {
+                List<Index> backingIndices = dataStream.getIndices();
+                // we'll look at the current write index in the next run if it's rolled over (and not the write index anymore)
+                for (int i = 0; i < backingIndices.size() - 1; i++) {
+                    IndexMetadata backingIndex = state.metadata().index(backingIndices.get(i));
+                    if (backingIndex == null || isManagedByDLM(dataStream, backingIndex) == false) {
+                        continue;
+                    }
+
+                    TimeValue indexLifecycleDate = getCreationOrRolloverDate(dataStream.getName(), backingIndex);
+
+                    long nowMillis = nowSupplier.getAsLong();
+                    if (nowMillis >= indexLifecycleDate.getMillis() + retention.getMillis()) {
+                        // there's an opportunity here to batch the delete requests (i.e. delete 100 indices / request)
+                        // let's start simple and reevaluate
+                        DeleteIndexRequest deleteRequest = new DeleteIndexRequest(backingIndex.getIndex().getName()).masterNodeTimeout(
+                            TimeValue.MAX_VALUE
+                        );
+
+                        // time to delete the index
+                        transportActionsDeduplicator.executeOnce(
+                            deleteRequest,
+                            ActionListener.noop(),
```

Comment:
Same question here about logging, I think it would be even more beneficial to log about DLM deleting indices (perhaps even at the `INFO` level, since it's not likely to happen as often and we should be explicit about it)